sdks/java/io/solr/src/main/java/org/apache/beam/sdk/io/solr/SolrIO.java

```diff
@@ -0,0 +1,684 @@
+              @Override public Long call() throws Exception {
+                return getEstimatedSizeOfShard(ReplicaInfo.create(replica));
+              }
+            });
+            futures.add(future);
+          }
+          for (Future<Long> future : futures) {
+            try {
+              sizeInBytes += future.get();
+            } catch (InterruptedException e) {
+              Thread.currentThread().interrupt();
+              throw new IOException(e);
+            } catch (ExecutionException e) {
+              throw new IOException("Can not estimate size of shard", e.getCause());
+            }
+          }
+        } finally {
+          executor.shutdownNow();
+        }
+
+        if (slices.size() <= 100) {
+          return sizeInBytes;
+        }
+        return (sizeInBytes / 100) * slices.size();
+      }
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      spec.populateDisplayData(builder);
+      if (replica != null) {
+        builder.addIfNotNull(DisplayData.item("shardUrl", replica.coreUrl()));
+      }
+    }
+
+    @Override
+    public BoundedReader<SolrDocument> createReader(PipelineOptions options)
+        throws IOException {
+      return new BoundedSolrReader(this);
+    }
+
+    @Override
+    public void validate() {
+      spec.validate(null);
+    }
+
+    @Override
+    public Coder<SolrDocument> getDefaultOutputCoder() {
+      return SolrDocumentCoder.of();
+    }
+  }
+
+  private static class BoundedSolrReader extends BoundedSource.BoundedReader<SolrDocument> {
+
+    private final BoundedSolrSource source;
+
+    private AuthorizedSolrClient solrClient;
+    private SolrDocument current;
+    private String cursorMark;
+    private Iterator<SolrDocument> batchIterator;
+    private boolean done;
+
+    private BoundedSolrReader(BoundedSolrSource source) {
+      this.source = source;
+      this.cursorMark = CursorMarkParams.CURSOR_MARK_START;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      if (source.replica != null) {
+        solrClient = source.spec.getConnectionConfiguration()
+            .createClient(source.replica.coreUrl());
+      } else {
+        solrClient = source.spec.getConnectionConfiguration()
+            .createClient();
+      }
+
+      SolrQuery solrParams = getQueryParams(source);
+      try {
+        QueryResponse response = solrClient.query(solrParams);
+        updateCursorMark(response);
+        return readNextBatchAndReturnFirstDocument(response);
+      } catch (SolrServerException e) {
+        throw new IOException(e);
+      }
+    }
+
+    private SolrQuery getQueryParams(BoundedSolrSource source) {
+      String query = source.spec.getQuery();
+      if (query == null) {
+        query = "*:*";
+      }
+      SolrQuery solrQuery = new SolrQuery(query);
+      solrQuery.set(CursorMarkParams.CURSOR_MARK_PARAM, cursorMark);
+      solrQuery.addSort("id", SolrQuery.ORDER.asc);
```

Comment:
By using cursor, the result must sort by uniqueField ( in 99% case it will "id" field, but I will modify solrio to grab the uniqueField's name from collection )